File: Java/org/oobench/exceptions/ExceptionsBenchmark.java
package org.oobench.exceptions;

import org.oobench.common.*;

public abstract class ExceptionsBenchmark extends AbstractBenchmark {

    public int getMajorNumber() {
        return 3;
    }

    public abstract void except(int count);
    public abstract void exceptDeep(int count);

    public void test(int count, String description) {
        count = getCountWithDefault(count);
        System.out.println("*** Testing " + description);

        for (int c = 0; c < 10; c++) {
            if (c == 9) {
                beginAction(1, "except", count);
            }
            except(count);
            if (c == 9) {
                endAction();
                beginAction(2, "except deep", count);
            }
            exceptDeep(count);
            if (c == 9) {
                endAction();
            }
        }
    }

    public static void testExceptions(Class theClass, int count, 
            String description, String[] args) {
        try {
            ExceptionsBenchmark bench = 
                    (ExceptionsBenchmark)theClass.newInstance();
            bench.setArguments(args);
            bench.test(count, description);
        } catch (Exception e) {
            e.printStackTrace();
            return;
        }
    }
}

File: C++/Exceptions/ExceptionsBenchmark.cpp
#include "AbstractBenchmark.cpp"

namespace oobench {

    class ExceptionsBenchmark : public AbstractBenchmark {
    public:
        ExceptionsBenchmark(int argc, const char** argv) 
                : AbstractBenchmark(argc, argv) { 
        }

        virtual int getMajorNumber() const {
            return 3;
        }

        virtual void except(unsigned long items) = 0;
        virtual void exceptDeep(unsigned long items) = 0;

        virtual const char* getExceptionsComment() {
            return "";
        }

        void test(unsigned long items, const char* description) {
            long count = getCountWithDefault(items);
            std::cout << "*** Testing " << description << std::endl;

            beginAction(1, "except", count, getExceptionsComment());
            except(count);
            endAction();

            beginAction(2, "except deep", count, getExceptionsComment());
            exceptDeep(count);
            endAction();
        }
    };
}
